test/puppetlabs/http/client/async_unbuffered_test.clj

```diff
@@ -327,3 +328,167 @@
+                      unbuffered-stream-data (get client-metrics-data unbuffered-stream-id)]
+                  (is (= {:count 1 :metric-id metric-id}
+                         (select-keys metric-data [:metric-id :count])))
+                  (is (<= 1 (:mean metric-data)))
+                  (is (<= 1 (:aggregate metric-data)))
+
+                  (is (= {:count 1 :metric-id unbuffered-stream-id}
+                         (select-keys unbuffered-stream-data [:metric-id :count])))
+                  (is (<= 1000 (:mean unbuffered-stream-data)))
+                  (is (<= 1000 (:aggregate unbuffered-stream-data)))
+
+                  (is (> (:mean unbuffered-stream-data) (:mean metric-data))))))))))
+     (testing "metrics work for a failed request"
+       (try
+         (testwebserver/with-test-webserver-and-config
+          (blocking-handler data) port {:shutdown-timeout-seconds 1}
+          (let [metric-registry (MetricRegistry.)]
+            (with-open [client (async/create-client {:connect-timeout-milliseconds 100
+                                                     :socket-timeout-milliseconds 200}
+                                                    metric-registry)]
+              (let [response @(common/get client (str "http://localhost:" port "/hello") opts)
+                    {:keys [body error]} response]
+                (is (nil? error))
+                ;; Consume the body to get the exception
+                (is (thrown? SocketTimeoutException (slurp body))))
+              (let [client-metrics (common/get-client-metrics client)
+                    client-metrics-data (common/get-client-metrics-data client)
+                    metric-id (str "puppetlabs.http-client.http://localhost:" port "/hello.GET")
+                    unbuffered-stream-id (str metric-id ".unbuffered_stream")]
+                (is (= (set (list metric-id unbuffered-stream-id))
+                       (set (keys client-metrics))
+                       (set (keys client-metrics-data))))
+                (is (every? #(instance? Timer %) (vals client-metrics)))
+                (let [metric-data (get client-metrics-data metric-id)
+                      unbuffered-stream-data (get client-metrics-data unbuffered-stream-id)]
+                  (is (= {:count 1 :metric-id metric-id}
+                         (select-keys metric-data [:metric-id :count])))
+                  (is (<= 1 (:mean metric-data)))
+                  (is (<= 1 (:aggregate metric-data)))
+
+                  (is (= {:count 1 :metric-id unbuffered-stream-id}
+                         (select-keys unbuffered-stream-data [:metric-id :count])))
+                  (is (<= 1 (:mean unbuffered-stream-data)))
+                  (is (<= 1 (:aggregate unbuffered-stream-data)))
+
+                  (is (> (:mean unbuffered-stream-data) (:mean metric-data))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
+           ))))))
```

Comment:
These tests are kind of gross and repetitive, but I didn't really want to overload the `java-non-blocking-streaming` and `clojure-non-blocking-streaming` functions even more, and I didn't want to start down a rabbit hole of refactoring this whole namespace.